### PR TITLE
Add ragged buffer to ring of experts

### PR DESCRIPTION
**Pull request by RissyRan**
# Description

Add ragged buffer to ring of experts, b/510333475.

# Tests

* Expect the tests are green with added ones
* Before this change - baseline, step time: 17.814s, 1840 TPS, [logs](https://screenshot.googleplex.com/7cQB7tYoHEwxA3r)
* After this change - ragged_buffer = 1.5, step time: 17.528s, 1870 TPS, [logs](https://screenshot.googleplex.com/AGqx9a8GQrtuBPj)

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [x] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run end-to-end tests tests and provided workload links above if applicable.
- [x] I have made or will make corresponding changes to the doc if needed, including adding new documentation pages to the relevant Table of Contents (toctree directive) as explained in [our documentation](https://maxtext.readthedocs.io/en/latest/development.html#adding-new-documentation-files).

### CODE REVIEW

File: src/maxtext/layers/moe.py [hunk 1]
```diff
@@ -808,13 +824,35 @@ def permute(self, inputs, gate_logits, pre_bias_logits, use_custom_sort_vjp=True
...
       group_size = jnp.bincount(flatten_selected_experts, length=self.num_experts)
-      # Return the experts for each sorted input.
-    expert_indices = jnp.arange(self.num_experts)
-    sorted_experts = jnp.repeat(
-        expert_indices,
-        repeats=group_size,
-        total_repeat_length=math.prod(selected_experts.shape),
-    )
+
+    n = bsz_times_seq_len * self.num_experts_per_tok
```
Comment (gobbleturk) [inline src/maxtext/layers/moe.py]: nit: use more descriptive var name than "n" - maybe num_tokens

File: src/maxtext/layers/moe.py [hunk 2]
```diff
@@ -808,13 +824,35 @@ def permute(self, inputs, gate_logits, pre_bias_logits, use_custom_sort_vjp=True
...
+    is_buffering = use_ragged_in_permute and buffer_size is not None and buffer_size < n
```
Comment (gobbleturk) [inline src/maxtext/layers/moe.py]: nit: would rename to using_truncated_buffer or using_small_buffer, is_buffering sounds like something is about to happen that hasn't happened yet

File: src/maxtext/configs/types.py [hunk 3]
```diff
@@ -2409,12 +2409,6 @@ def _validate_check_vma_is_supported(self):
           f"Found other ICI axes enabled: {active}."
       )
 
-  def validate_ragged_buffer_factor(self):
-    if self.ragged_buffer_factor <= 0:
-      return  # Nothing to validate if not using ragged buffer factor
-    if self.use_ring_of_experts:
-      raise ValueError("Currently we only support ragged buffer factor with ragged a2a approach.")
```
Comment (NuojCheng) [inline src/maxtext/configs/types.py:2421]: Is ragged buffer factor supported for ring EP when ragged sort is not used?

File: src/maxtext/layers/moe.py [hunk 4]
```diff
@@ -1566,14 +1622,28 @@ def wrapper(x, logits, pre_bias_logits, w0, w1, wo, w0_bias, w1_bias, wo_bias, r
...
       num_experts_per_shard = self.config.num_experts // num_ep
-      if self.config.use_ragged_sort and self.config.use_ring_of_experts:
-        experts_start = route_metadata.expert_shard_id * num_experts_per_shard
-      else:
-        experts_start = 0
 
-      gmm_fn = functools.partial(
-          gmm, group_sizes=routing.group_sizes, expert_assignments=routing.selected_experts, group_offset=experts_start
-      )
+      is_buffering = self.config.use_ring_of_experts and x.shape[0] < routing.sorted_selected_experts.shape[0]
```
Comment (gobbleturk) [inline src/maxtext/layers/moe.py]: is there no way to pass to this is_buffering into sort - we have to calculate this twice?

File: tests/unit/moe_test.py [hunk 5]
```diff
@@ -715,11 +716,22 @@ def loss_fn(params, x):
   def test_ragged_sort_loss_and_grad_ring_of_experts(self):
     self._run_ragged_sort_loss_and_grad(use_ring_of_experts=True)
 
+  @pytest.mark.tpu_only
+  @pytest.mark.skip_on_tpu7x
+  def test_ragged_sort_loss_and_grad_ring_of_experts_ragged_buffer(self):
+    # no dropping is expected
```
Comment (gobbleturk) [inline tests/unit/moe_test.py]: why is no dropping expected here?

**Review by gobbleturk — APPROVED** [review]

File: src/maxtext/layers/moe.py (hunk 1 above)
Comment (RissyRan) [reply src/maxtext/layers/moe.py]: ok, will update.

File: src/maxtext/layers/moe.py (hunk 2 above)
Comment (RissyRan) [reply src/maxtext/layers/moe.py]: ok, sg! Will update to `use_truncated_buffer`

File: src/maxtext/configs/types.py (hunk 3 above)
Comment (RissyRan) [reply src/maxtext/configs/types.py:2421]: Good call! Let me cross check.

File: tests/unit/moe_test.py (hunk 5 above)
Comment (RissyRan) [reply tests/unit/moe_test.py]: I see inputs are initialized randomly so theoretically it should be balanced distributed. I will remove this to avoid confusion.

File: src/maxtext/configs/types.py (hunk 3 above)
Comment (RissyRan) [reply src/maxtext/configs/types.py:2421]: Update the validation and tests accordingly. In the previous case, it just directly fall back to the -1.0 case for non ragged sort with ring of experts.

File: src/maxtext/layers/moe.py (hunk 4 above)
Comment (RissyRan) [reply src/maxtext/layers/moe.py]: Good call! Updated to `if routing.local_group_sizes is not None`, and re-tested in cloud [run](https://cloudlogging.app.goo.gl/rHduZ6sw7fxuso9m8) as expected.

**Review by NuojCheng — APPROVED** [review]
Thank you Ran for onboarding this awesome feature!
